draft-ietf-quic-transport.md

```diff
@@ -884,6 +884,54 @@ provide an interface to QUIC to tell it about its buffering limits so that there
 is not excessive buffering at multiple layers.
 
 
+## Flow Control Deadlocks {#flow-control-deadlocks}
+
+Flow control provides a means of managing access to the limited buffers
+endpoints have for incoming data.  This mechanism limits the amount of data that
+can be in buffers in endpoints or in transit on the network.  However, there are
+several ways in which limits can produce conditions that can cause a connection
+to either perform suboptimally or deadlock.
+
+Implementations need to be aware of the possibility of deadlocking and that some
+conditions that lead to deadlocking cannot be completely avoided in a protocol
+design.
+
+Large messages can produce deadlocking if the recipient does not process the
+message incrementally.  If the message is larger than flow control credit
+available and the recipient does not release additional flow control credit, a
+deadlock can occur.  This is possible even where stream flow control limits are
+not reached because connection flow control limits can be consumed by other
+streams.
+
+A common implementation technique has flow control credit extended as data is
+read from receive buffers.  In that setting, a length-prefixed message format
+makes it easier for a recipient to leave data unread and thereby withhold flow
+control credit.  If flow control limits prevent the remainder of a message from
+being sent, a deadlock will result.  However, a length prefix might enable the
+detection of this sort of deadlocks.  Where protocols have messages that might
+be processed as a single unit, reserving flow control credit for the entire
```

Comment:
Connection level flow control can't be reduced, so are you suggesting it not be further increased until the message arrives?  I haven't implemented this sort of logic, but it seems very tricky to get right and done incorrectly, could result in deadlocks.

Reply:
Concrete example:  You can't consume less than a full QPACK instruction on the encoder stream.  If you read far enough to see that you have a 64KB auth header coming at you, you might logically not continue reading until you've received those 64KB.  If there's not sufficient flow control for the sender to complete sending, nothing gets read and you deadlock.

That might be because the connection flow control never gets increased, or it might be because it got increased but allocated to other streams that have data to send.

Reply:
> Concrete example: You can't consume less than a full QPACK instruction on the encoder stream.

Yes, you can. See [lsqpack_dec_enc_in()](https://github.com/litespeedtech/ls-qpack/blob/5f17c82e8c0f3b2b2faa8984256ff355bc5dc657/src/lsqpack.c#L4186)

Reply:
I think he's just giving an example of an implementation that cannot handle this.  You are correct that the protocol does not mandate this specific behavior.